File: packages/sdk-python/src/nullspend/_loop_detector.py
"""Sliding-window loop detection for repeated LLM calls."""
from __future__ import annotations

import hashlib
import threading
import time
from collections import defaultdict
from typing import NamedTuple


class LoopCheck(NamedTuple):
    is_loop: bool
    is_warning: bool
    call_count: int
    detection_type: str = "per_key"  # "per_key" or "aggregate"


class LoopDetector:
    """Detects repeated calls to the same model+content within a time window.

    Thread-safe. One instance per tracked client.
    Time-pruned lists, bounded by window expiry.
    """

    def __init__(
        self,
        max_calls: int = 50,
        window_seconds: float = 60.0,
        aggregate_max_keys: int = 5,
        aggregate_min_repeats: int = 3,
        warning_ratio: float = 0.8,
    ):
        if max_calls < 0:
            raise ValueError(f"max_calls must be >= 0 (got {max_calls})")
        if window_seconds <= 0:
            raise ValueError(f"window_seconds must be > 0 (got {window_seconds})")
        if aggregate_max_keys < 0:
            raise ValueError(f"aggregate_max_keys must be >= 0 (got {aggregate_max_keys})")
        self._max_calls = max_calls
        self._window = window_seconds
        self._aggregate_max_keys = aggregate_max_keys
        self._aggregate_min_repeats = aggregate_min_repeats
        self._warning_ratio = warning_ratio
        # key -> list of (timestamp, content_hash) — time-pruned on each check
        self._call_log: dict[str, list[tuple[float, str]]] = defaultdict(list)
        self._warnings_fired: set[str] = set()
        self._lock = threading.Lock()

    @staticmethod
    def content_hash(body: bytes | str | None) -> str:
        """Hash full request body to 8 hex chars. Cap at 8KB for large payloads."""
        if not body:
            return "empty"
        if isinstance(body, str):
            raw = body[:8192].encode("utf-8")
        else:
            raw = body[:8192]
        return hashlib.sha256(raw).hexdigest()[:8]

    def check(self, key: str, content_hash: str) -> LoopCheck:
        """Record a call and check for loops.

        Returns LoopCheck(is_loop, is_warning, call_count).
        """
        now = time.monotonic()
        cutoff = now - self._window
        warning_count = int(self._max_calls * self._warning_ratio)

        with self._lock:
            # Per-key check: prune old entries, append new, then cap to bound
            # memory under burst load with many distinct content hashes (SEC-2).
            # Cap at 10× max_calls — enough headroom for legitimate detection
            # (counting by hash) while preventing unbounded growth.
            entries = self._call_log[key]
            self._call_log[key] = [(t, h) for t, h in entries if t > cutoff]
            entries = self._call_log[key]
            entries.append((now, content_hash))
            entry_cap = max(self._max_calls * 10, self._max_calls)
            if len(entries) > entry_cap:
                del entries[0:len(entries) - entry_cap]

            # Count matching content hash
            count = sum(1 for _, h in entries if h == content_hash)

            # Per-key loop
            is_loop = count >= self._max_calls
            detection_type = "per_key"

            # Aggregate: count distinct keys with 3+ same-content repeats.
            # Prune all keys lazily here to avoid counting stale entries
            # from keys that haven't been checked recently (BUG-1 fix).
            if not is_loop:
                qualifying_keys = 0
                stale_keys: list[str] = []
                for k, ents in list(self._call_log.items()):
                    if k != key:  # Current key already pruned above
                        ents = [(t, h) for t, h in ents if t > cutoff]
                        self._call_log[k] = ents
                    if not ents:
                        stale_keys.append(k)
                        continue
                    hash_counts: dict[str, int] = {}
                    for _, h in ents:
                        hash_counts[h] = hash_counts.get(h, 0) + 1
                    if any(c >= self._aggregate_min_repeats for c in hash_counts.values()):
                        qualifying_keys += 1
                # Clean up fully-expired keys to prevent dict growth
                for k in stale_keys:
                    del self._call_log[k]
                if qualifying_keys >= self._aggregate_max_keys:
                    is_loop = True
                    count = qualifying_keys
                    detection_type = "aggregate"

            # Warning (fires once per window per composite key, uses >= not ==)
            composite = f"{key}:{content_hash}"
            is_warning = False
            if count >= warning_count and composite not in self._warnings_fired:
                is_warning = True
                self._warnings_fired.add(composite)
            # Reset warning if dropped below threshold
            if count < warning_count:
                self._warnings_fired.discard(composite)

            return LoopCheck(is_loop=is_loop, is_warning=is_warning, call_count=count, detection_type=detection_type)

    def reset(self, key: str | None = None) -> None:
        """Reset call history."""
        with self._lock:
            if key is None:
                self._call_log.clear()
                self._warnings_fired.clear()
            else:
                self._call_log.pop(key, None)
                self._warnings_fired = {
                    k for k in self._warnings_fired if not k.startswith(f"{key}:")
                }

File: packages/sdk/src/loop-detector.ts
/**
 * Sliding-window loop detection for repeated LLM calls.
 *
 * Mirror of Python implementation. Uses Map<string, Array<{ts, hash}>>.
 * No lock needed (single-threaded Node.js).
 */

export interface LoopCheck {
  isLoop: boolean;
  isWarning: boolean;
  callCount: number;
  detectionType: "per_key" | "aggregate";
}

export interface LoopDetectionOptions {
  maxCalls?: number;      // default 50
  windowSeconds?: number; // default 60
  aggregateMaxKeys?: number; // default 5
}

export class LoopDetector {
  private readonly _maxCalls: number;
  private readonly _window: number; // seconds
  private readonly _aggregateMaxKeys: number;
  private readonly _aggregateMinRepeats = 3;
  private readonly _warningRatio = 0.8;

  /** key → list of { ts (ms monotonic), hash } */
  private readonly _callLog = new Map<string, Array<{ ts: number; hash: string }>>();
  private readonly _warningsFired = new Set<string>();

  constructor(options?: LoopDetectionOptions) {
    const maxCalls = options?.maxCalls ?? 50;
    const windowSeconds = options?.windowSeconds ?? 60;
    const aggregateMaxKeys = options?.aggregateMaxKeys ?? 5;

    if (maxCalls < 0) throw new RangeError(`maxCalls must be >= 0 (got ${maxCalls})`);
    if (windowSeconds <= 0) throw new RangeError(`windowSeconds must be > 0 (got ${windowSeconds})`);
    if (aggregateMaxKeys < 0) throw new RangeError(`aggregateMaxKeys must be >= 0 (got ${aggregateMaxKeys})`);

    this._maxCalls = maxCalls;
    this._window = windowSeconds;
    this._aggregateMaxKeys = aggregateMaxKeys;
  }

  /**
   * Hash full request body to 8 hex chars. Cap at 8KB for large payloads.
   * Uses SHA-256 when available (Web Crypto), falls back to simple hash.
   */
  static contentHashSync(body: string | null | undefined): string {
    if (!body) return "empty";
    const slice = body.slice(0, 8192);
    // Simple but deterministic hash for synchronous path (FNV-1a 32-bit → 8 hex)
    let hash = 0x811c9dc5;
    for (let i = 0; i < slice.length; i++) {
      hash ^= slice.charCodeAt(i);
      hash = Math.imul(hash, 0x01000193);
    }
    // Convert to unsigned 32-bit, then 8 hex chars
    return (hash >>> 0).toString(16).padStart(8, "0");
  }

  /**
   * Async SHA-256 content hash matching the proxy implementation.
   * Use this when Web Crypto is available and you can await.
   */
  static async contentHash(body: string | null | undefined): Promise<string> {
    if (!body) return "empty";
    const slice = body.slice(0, 8192);
    if (typeof globalThis.crypto?.subtle?.digest === "function") {
      const buf = await crypto.subtle.digest("SHA-256", new TextEncoder().encode(slice));
      const arr = new Uint8Array(buf);
      return Array.from(arr.slice(0, 4), (b) => b.toString(16).padStart(2, "0")).join("");
    }
    // Fallback to sync hash
    return LoopDetector.contentHashSync(body);
  }

  /**
   * Record a call and check for loops.
   */
  check(key: string, contentHash: string): LoopCheck {
    const now = performance.now();
    const cutoffMs = now - this._window * 1000;
    const warningCount = Math.floor(this._maxCalls * this._warningRatio);
    let detectionType: "per_key" | "aggregate" = "per_key";

    // Per-key: prune old entries, append new, and cap to bound memory under
    // burst load with many distinct content hashes (SEC-2). Cap at 10× maxCalls
    // — enough headroom for legitimate detection (counting by hash, see below)
    // while preventing unbounded growth from a high-QPS attacker.
    let entries = this._callLog.get(key) ?? [];
    entries = entries.filter((e) => e.ts > cutoffMs);
    entries.push({ ts: now, hash: contentHash });
    const entryCap = Math.max(this._maxCalls * 10, this._maxCalls);
    if (entries.length > entryCap) {
      entries = entries.slice(entries.length - entryCap);
    }
    this._callLog.set(key, entries);

    // Count matching content hash
    let count = 0;
    for (const e of entries) {
      if (e.hash === contentHash) count++;
    }

    // Per-key loop
    let isLoop = count >= this._maxCalls;

    // Aggregate: count distinct keys with 3+ same-content repeats
    // Prune all keys lazily to avoid counting stale entries
    if (!isLoop) {
      let qualifyingKeys = 0;
      const staleKeys: string[] = [];
      for (const [k, ents] of this._callLog) {
        let filtered = ents;
        if (k !== key) {
          filtered = ents.filter((e) => e.ts > cutoffMs);
          this._callLog.set(k, filtered);
        }
        if (filtered.length === 0) {
          staleKeys.push(k);
          continue;
        }
        const hashCounts = new Map<string, number>();
        for (const e of filtered) {
          hashCounts.set(e.hash, (hashCounts.get(e.hash) ?? 0) + 1);
        }
        for (const c of hashCounts.values()) {
          if (c >= this._aggregateMinRepeats) {
            qualifyingKeys++;
            break;
          }
        }
      }
      for (const k of staleKeys) {
        this._callLog.delete(k);
      }
      if (qualifyingKeys >= this._aggregateMaxKeys) {
        isLoop = true;
        count = qualifyingKeys;
        detectionType = "aggregate";
      }
    }

    // Warning (fires once per window per composite key)
    const composite = `${key}:${contentHash}`;
    let isWarning = false;
    if (count >= warningCount && !this._warningsFired.has(composite)) {
      isWarning = true;
      this._warningsFired.add(composite);
    }
    if (count < warningCount) {
      this._warningsFired.delete(composite);
    }

    return { isLoop, isWarning, callCount: count, detectionType };
  }

  /** Reset call history (all keys or a specific key). */
  reset(key?: string): void {
    if (key === undefined) {
      this._callLog.clear();
      this._warningsFired.clear();
    } else {
      this._callLog.delete(key);
      const prefix = `${key}:`;
      for (const k of this._warningsFired) {
        if (k.startsWith(prefix)) this._warningsFired.delete(k);
      }
    }
  }

  get maxCalls(): number { return this._maxCalls; }
  get windowSeconds(): number { return this._window; }
}
